app/src/common/shared/org/mozilla/vrbrowser/ui/keyboards/ChineseZhuyinKeyboard.java

```diff
@@ -0,0 +1,411 @@
+                        String key = getString(cursor, 0);
+                        String word = getString(cursor, 1);
+                        addToKeyMap(aKey, key, word);
+                        --limit;
+                    } while (limit >= 0 && cursor.moveToNext());
+                }
+            } catch (Exception e) {
+                Log.e(LOGTAG, "Querying Zhuyin db failed");
+            }
+        }
+
+        if (limit <= 0) {
+            return;
+        }
+
+        // Query phrase
+        reader = mPhraseDB.getReadableDatabase();
+        sqliteArgs[0] = transCode + '%';
+        sqliteArgs[1] = "" + limit;
+        try (Cursor cursor = reader.rawQuery("SELECT code, word FROM phrases_" + transCode.substring(0, 2)
+                + " where code like ? group by word ORDER BY frequency DESC LIMIT ?", sqliteArgs)) {
+            if (cursor.moveToFirst()) {
+                do {
+                    String key = getString(cursor, 0);
+                    String word = getString(cursor, 1);
+                    addToKeyMap(aKey, key, word);
+                    --limit;
+                } while (limit >= 0 && cursor.moveToNext());
+            }
+        } catch (Exception e) {
+            Log.e(LOGTAG, "Querying Zhuyin db failed");
+        }
+    }
+
+    private void addToKeyMap(String aKey, String aCode, String aDisplays) {
+        if (aKey == null || aKey.isEmpty()) {
+            Log.e(LOGTAG, "Zhuyin key is null");
+            return;
+        }
+        if (aCode == null || aCode.isEmpty()) {
+            Log.e(LOGTAG, "Zhuyin code is null");
+            return;
+        }
+        KeyMap keyMap = mKeymaps.get(aKey);
+        if (keyMap == null) {
+            keyMap = new KeyMap();
+            mKeymaps.put(aKey, keyMap);
+        }
+
+        if (aDisplays != null && !aDisplays.isEmpty()) {
+            String[] displayList = aDisplays.split("\\|");
+            for (String display: displayList) {
+                keyMap.displays.add(new Words(syllableCount(aCode), aCode, display));
+            }
+        }
+    }
+
+    private void addKeyCode(String aKey, String aCode, String aDisplay) {
+        mKeyCodes.put(aKey, new Words(syllableCount(aCode), aCode, aDisplay));
+    }
+
+    private int syllableCount(String aCode) {
+        if (aCode == null) {
+            return 0;
+        }
+        aCode = aCode.trim();
+        if (aCode.isEmpty()) {
+            return 0;
+        }
+
+        return (int)aCode.chars().filter(ch -> ch == ' ').count() + 1;
+    }
+
+    private String getString(Cursor aCursor, int aIndex) {
+        if (aCursor.isNull(aIndex)) {
+            return null;
+        }
+        return aCursor.getString(aIndex);
+    }
+
+
+    class KeyMap {
+        ArrayList<Words> displays = new ArrayList<>();
+    }
+
+
+    class DBWordHelper extends SQLiteAssetHelper {
+        private static final String DATABASE_NAME = "zhuyin_words.db";
+        private static final int DATABASE_VERSION = 1;
```

Comment:
@cvan If we want to update the database, we can follow the instruction [here](https://github.com/jgilfelt/android-sqlite-asset-helper#database-upgrades). But, our dictionary only needs to be read for users instead of writing. So, I guess we can handle this at our internal process.